idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/proofs/intentions/coercion.kt

```diff
@@ -0,0 +1,80 @@
+package arrow.meta.ide.plugins.proofs.intentions
+
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.log.Log
+import arrow.meta.log.invoke
+import arrow.meta.phases.CompilerContext
+import arrow.meta.phases.ExtensionPhase
+import arrow.meta.phases.resolve.baseLineTypeChecker
+import arrow.meta.plugins.proofs.phases.coerceProof
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.debugger.sequence.psi.resolveType
+import org.jetbrains.kotlin.jsr223.KotlinJsr223StandardScriptEngineFactory4Idea
+import org.jetbrains.kotlin.nj2k.postProcessing.type
+import org.jetbrains.kotlin.psi.KtDotQualifiedExpression
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.module
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+
+fun IdeMetaPlugin.coercionIntention(): ExtensionPhase =
+  addIntention(
+    text = "Make coercion explicit",
+    kClass = KtElement::class.java,
+    isApplicableTo = { ktCall: KtElement, caretOffset: Int ->
+      when (ktCall) {
+        is KtDotQualifiedExpression -> ktCall.isDotQualifiedExpressionCoerced()
+        is KtProperty -> ktCall.isCoerced()
+        else -> false
+      }
+    },
+    applyTo = { ktCall, editor ->
+      // replace with proof previously found
+    }
+  )
+
+private fun KtDotQualifiedExpression.isDotQualifiedExpressionCoerced(): Boolean {
+  val bindingContext: BindingContext = analyze(bodyResolveMode = BodyResolveMode.FULL)
+  val calls = bindingContext.getSliceContents(BindingContext.RESOLVED_CALL)
+  return calls.filter { (call, resolvedCall) -> resolvedCall != null }
+    .any { (call, resolvedCall) ->
+      selectorExpression?.let { selectorExpression ->
+        val (type, superType) = receiverExpression.resolveType() to selectorExpression.resolveType()
+        val isSubtypeOf = baseLineTypeChecker.isSubtypeOf(type, superType)
+        val compilerContext = CompilerContext(project = call.callElement.project, eval = {
+          KotlinJsr223StandardScriptEngineFactory4Idea().scriptEngine.eval(it)
```

Comment:
How does this work? So you create some sort of default script engine for Idea that you can use to evaluate the expression to get what exactly?

Reply:
I'll redirect that question to the masters @raulraja @i-walker  :)

Reply:
This is the IDE impl of the refine types predicate. 

The Kotlin plugin provides a script engine factory that connects to the Kotlin compiler daemon in the user IDE and uses it to evaluate the predicate against values such as `TwitterHandle("@admin")` which in this case would fail to show at compile time in the editor that that is not OK. 

Depending on whether we are running the CLI compiler or the IDEA embedded compiler we need to use different implementations.

Then we will register an intention to ask the user to use `TwitterHandle.from(value)` which is a safe API for a runtime value or `TwitterHandle("@whatever")` which is a valid identifier.